lib/Analysis/SchemeInfoAnalysis/SchemeInfoAnalysis.h

```diff
@@ -0,0 +1,248 @@
+    assert(isInitialized() && "NatureOfComputation not initialized");
+    //TODO: define what happens when all are equal
+    int maxCount = numBoolOps;
+    StringRef attributeName = numBoolOpsAttrName;
+
+    if (numBitOps > maxCount) {
+        maxCount = numBitOps;
+        attributeName = numBitOpsAttrName;
+    }
+    if (numIntArithOps > maxCount) {
+        maxCount = numIntArithOps;
+        attributeName = numIntArithOpsAttrName;
+    }
+    if (numRealArithOps > maxCount) {
+        maxCount = numRealArithOps;
+        attributeName = numRealArithOpsAttrName;
+    }
+    if (numCmpOps > maxCount) {
+        maxCount = numCmpOps;
+        attributeName = numCmpOpsAttrName;
+    }
+    if (numNonLinOps > maxCount) {
+        maxCount = numNonLinOps;
+        attributeName = numNonLinOpsAttrName;
+    }
+
+    return attributeName;
+}
+
+int getDominantComputationCount() const {
+    assert(isInitialized() && "NatureOfComputation not initialized");
+    
+    int maxCount = numBoolOps;
+
+    if (numBitOps > maxCount) {
+        maxCount = numBitOps;
+    }
+    if (numIntArithOps > maxCount) {
+        maxCount = numIntArithOps;
+    }
+    if (numRealArithOps > maxCount) {
+        maxCount = numRealArithOps;
+    }
+    if (numCmpOps > maxCount) {
+        maxCount = numCmpOps;
+    }
+    if (numNonLinOps > maxCount) {
+        maxCount = numNonLinOps;
+    }
+
+    return maxCount; // Returns the highest count after all comparisons
+}
+
+  static NatureOfComputation max(const NatureOfComputation &lhs,
+                                 const NatureOfComputation &rhs) {
+    return NatureOfComputation(
+        std::max(lhs.numBoolOps, rhs.numBoolOps),
+        std::max(lhs.numBitOps, rhs.numBitOps),
+        std::max(lhs.numIntArithOps, rhs.numIntArithOps),
+        std::max(lhs.numRealArithOps, rhs.numRealArithOps),
+        std::max(lhs.numCmpOps, rhs.numCmpOps),
+        std::max(lhs.numNonLinOps, rhs.numNonLinOps));
+  }
+
+  static NatureOfComputation join(const NatureOfComputation &lhs,
+                                  const NatureOfComputation &rhs) {
+    if (!lhs.isInitialized()) {
+      return rhs;
+    }
+
+    if (!rhs.isInitialized()) {
+      return lhs;
+    }
+
+    return NatureOfComputation::max(lhs, rhs);
+  }
+
+  void print(llvm::raw_ostream &os) const {
+    if (isInitialized()) {
+      os << "NatureOfComputation(numBoolOps=" << numBoolOps
+         << "; numBitOps=" << numBitOps << "; numIntArithOps=" << numIntArithOps
+         << "; numRealArithOps=" << numRealArithOps << "; numCmpOps=" << numCmpOps
+         << "; numNonLinOps=" << numNonLinOps << ")";
+    } else {
+      os << "NatureOfComputation(uninitialized)";
+    }
+  }
+
+  friend llvm::raw_ostream &operator<<(llvm::raw_ostream &os,
+                                       const NatureOfComputation &state) {
+    state.print(os);
+    return os;
+  }
+
+ private:
+  bool initialized;
+  int numBoolOps;
+  int numBitOps;
+  int numIntArithOps;
```

Comment:
in a scheme selection, i would think that if arith ops have smaller bit width (2-8 bits), the cost of evaluation in CGGI like schemes will be lower than high bit width arithmetic (8+ bits). i think this info doesn't synthesize this but maybe the cost model can take this into consideration?